### DISCUSSION TITLE
Cannot specify full boundary conditions in biharmonic equations without periodic boundary conditions

### DISCUSSION THREAD

**Discussion by tranqui** [zwicker-group/py-pde#310] [Q&A] [answered]
I am trying to solve 4th order ODEs containing a biharmonic operator. The documentation describes solving the [Kuramoto-Sivashinsky](https://py-pde.readthedocs.io/en/latest/examples_gallery/pde_custom_numba.html#sphx-glr-examples-gallery-pde-custom-numba-py) equation and the [Cahn-Hilliard](https://py-pde.readthedocs.io/en/latest/packages/pde.pdes.cahn_hilliard.html) equation.

However, there is no option to specify enough boundary conditions to solve a 4th order ODE. For a linear ODE we need 4 and this is generally true for nonlinear ODEs also. With Neumann BCs I expect to be able to specify values and derivatives at the same point, with e.g. something like for a 2d system:
`bcs = [{'value': -1, 'derivative': 0}, {'value': 1, 'derivative': 0}]`
However, this always throws an error saying this feature is not supported. So I am curious how 4th order ODEs are listed as examples. Ultimately I'm interested in something similar to the Cahn-Hilliard equation so look to that as an example.

Looking at the implementation of the Cahn-Hilliard equation, it looks like this would only work with periodic boundary conditions because this implicitly sets values and all derivatives. That being said, I am never able to get convergent results for e.g. the binodal so I suspect either I am using py-pde wrong or there is a problem with the underlying solver. Below is a minimal working example that illustrates this where I try to obtain the binodal. I start with two interfaces around x=+-0.5 to respect periodic boundaries at x=+-1. The solver quickly fails with a "Field was not finite" error:
```
from pde import CahnHilliardPDE, ScalarField, CartesianGrid
grid = CartesianGrid([[-1, 1], [-1, 1]], [1000,50])
state = ScalarField.from_expression(grid, 'tanh(cos(pi*x)/0.1)')
eq = CahnHilliardPDE()
result = eq.solve(state, 1, dt=1e-6, method='implicit')
result.plot()
```

My expected behaviour is that it should converge to tighten up the interfaces with the correct interface width (rather than the dummy value of 0.1 I gave as an IC).

Is this a bug, or am I using it wrong? If the latter could you please advise me on how I should use the Cahn Hilliard (and similar) solvers? Thank you.

**Comment by david-zwicker** [comment, marked as answer]
You're asking multiple questions instead of describing a particular issue with the code, so I converted it into a discussion item. I'll try to address the questions I could find:

- A biharmonic operator is basically two nested Laplace operators. Each Laplace operator requires a boundary condition at each boundary point in `py-pde`. The `CahnHilliardPDE` class allows setting these conditions using the `bc_c` and `bc_mu` flag. When using the `PDE` class, you can set specific boundary conditions for specific operators using the `bc_ops` argument; see [the documentation](https://py-pde.readthedocs.io/en/latest/packages/pde.pdes.pde.html#pde.pdes.pde.PDE).
- I don't know what you tried to achieve when setting the boundary condition `bcs = [{'value': -1, 'derivative': 0}, {'value': 1, 'derivative': 0}]`, but I'm not surprised that you got an error.
- Your code for the Cahn-Hilliard equation will likely not converge because you discretization length is orders of magnitude smaller than the interface length. A working example reads:
```python
from pde import CahnHilliardPDE, ScalarField, CartesianGrid
grid = CartesianGrid([[-10, 10], [-10, 10]], 30)
state = ScalarField.from_expression(grid, 'tanh(cos(0.1 * pi*x)/0.1)')
eq = CahnHilliardPDE()
result = eq.solve(state, 10, dt=1e-6, adaptive=True)
result.plot()
```
I hope this gets you on the right track!

**Reply by tranqui** [reply]
I take your point about this being more of a discussion item. Thank you for the fast reply, it was very helpful.

Regarding what I was trying to achieve with the boundary conditions, I was trying to fix the value and derivative of the x-component of the scalar field (let's call it u(x)) at each boundary. E.g. u(-1) = -1, u'(-1) = 0, u(1) = 1 and u'(1) = 0. I tried a few different ways of specifying those boundary conditions, but it doesn't look like that functionality exists?

Thank you for correcting the Cahn-Hilliard code, that's definitely got me on track. It was quite silly of me to be using the completely wrong length scale XD

**Reply by david-zwicker** [reply]
How to implement boundary conditions not only depends on the fields, but also on the differential operators since they are the ones where boundary conditions are used. To understand what you're doing I would need to see the actual equations.

**Reply by tranqui** [reply]
Thank you for indulging me.

Okay, let me give a 1d example. Suppose our system was $u^{(4)}(x) = 0$ with boundary conditions $u(0) = u'(0) = u'(1) = 0$ and $u(1) = 1$. The analytic solution of this is $u(x) = 3x^2 - 2x^3$. This system might correspond to an asymptotically thin beam that can bend but not stretch. The field would then be interpreted as the deflection of said beam, and the BCs correspond to clamping the beam on each end.

**Reply by david-zwicker** [reply]
Thanks for the example. I now see what you're trying to do. I'm afraid that this might be currently difficult to implement with `py-pde` since the package focuses on the time evolution of field equations. In particular, the PDEs are typically agnostic to geometry and even dimension, whereas your equation seems to work in 1d only.

**Reply by tranqui** [reply]
I see, so it might not be possible to do what I'm after. The example I gave is simple it's true, but readily generalisable to higher dimensions. For example, it is a limiting case of the [Föppl-von Kármán equations](https://en.wikipedia.org/wiki/F%C3%B6ppl%E2%80%93von_K%C3%A1rm%C3%A1n_equations) which is a field theory describing deformations of elastic plates. The time-dependent form of these equations would exactly fit what you describe. They are 2-dimensional, and you must always specify 4 boundary conditions. These are notoriously difficult equations to work with however, so perhaps they are an unfair example to cite.

**Reply by david-zwicker** [reply]
Thank you for pointing out these equations. I agree that it would be nice to be able to simulate the equations in `py-pde`. I actually see no principle obstacle to do this if one can turn them into dynamical equations. In the simplest case, these dynamical equations can relax toward the stationary problem specified on the Wikipedia page. More challenging will be the implementation of the boundary conditions that you propose. I'm not sure how one could impose both the value and the derivative at a boundary point in any general way. We so far often dealt with imposing value and second derivative, or first derivative and second derivatives of fields. In such a case, there is a clear association with the two nested Laplace operators in the biharmonic equation.

**Reply by tranqui** [reply]
Curious. I don't know how py-pde is working under the hood. I can have a think about this if you point me to where the relevant parts of code where BCs are applied. Coming from more of a finite-element background, one generic way to impose the kinds of BCs I describe would be to introduce ghost points at the edges which automatically satisfy the BCs.

**Reply by david-zwicker** [reply]
Yes, this is exactly what we do, too. Under the hood, `py-pde` allocates arrays with dimension `N + 2` to represent a field with `N` support points (I'm considering the 1D case for simplicity). The indices `1, ..., N` correspond to the real field values, while the index `0` and `N - 1` are the respective virtual points. Applying boundary conditions to fields amounts to setting these virtual points based on the neighboring values. Using these ghost points, differential operators can then be evaluated in the entire domain (so for `i = 1, ..., N`). The biharmonic operator would currently be implemented as nested Laplacian, which implies setting virtual points for the actual field (to evaluate the first Laplacian) and then ghost points of the field containing the second derivatives (to apply the second Laplacian). I'm struggling how to implement the second part for your boundary conditions, but it might very well be possible!

**Reply by david-zwicker** [reply]
By the way, we have [a little document in the repository, which contains a bit more details on the boundary conditions](https://github.com/zwicker-group/py-pde/blob/master/docs/methods/boundary_discretization/boundary_discretization.pdf).

**Reply by tranqui** [reply]
Many thanks, that's a very helpful document! I think I see the problem. Normally in order to go to fourth order one has to introduce 2 ghost points on each boundary, so indices $\\{0, 1, N-2, N-1\\}$ would become the ghost points. Then when you evaluate the laplacian twice (or otherwise use a next-nearest neighbour stencil) you can still apply two BCs on the edge points. So I guess that would require modifying the grid class (if that's where ghost points are encoded?).

If the architecture is such that adding additional ghost points would be unwieldy, perhaps a hacky solution would be to switch to a non-central stencil at the boundaries when additional BCs are to be applied. Either a forward/backward or a mixed stencil with one ghost point would do the trick. That would still require next-nearest neighbours (in the inward direction) though to achieve second-order accuracy.

Equation (11) of the document linked above puzzles me. It looks like you're only applying first-order accurate first derivatives at the boundary, and not strictly taking the boundary value $f(0) = y_1$ but rather the mid-point between it and the virtual point. Why not simply continue the second-order stencil?

**Reply by tranqui** [reply]
Right, you apply BCs at the operator level so you have to precalculate the ghost points. So they're only ever known to first-order. Introducing additional ghost points would then potentially improve the overall order of the solver then?

**Reply by tranqui** [reply]
Though you could use $f(0) = \frac{2y_0 + y_{-1} + y_1}{4} + \mathcal{O}(\Delta x^2)$ and $f'(0) = \frac{y_1 - y_{-1}}{2\Delta x} + \mathcal{O}(\Delta x^2)$. It's straightforward to find similar expressions for higher-order ghost points - I can provide you with expressions/code if that would be helpful?

**Reply by david-zwicker** [reply]
Note that we do not have support points on the boundary but rather the boundary is half way between support points; see [the documentation](https://py-pde.readthedocs.io/en/latest/manual/mathematical_basics.html#spatial-discretization). This has many advantages and allows to express values and derivatives using Eq. 11 in the PDF document. I agree that a fourth-order derivative could in principle be implemented with a five-point stencil, but since this requires two virtual points in each direction, we didn't do this. Instead, we implement the biharmonic operator by applying the Laplacian twice, which lets us easily impose boundary conditions for `u` and `u''`. There might be a way to calculate the virtual point for `u''` from the boundary condition on `u'`, but this would probably also involve knowledge about the actual differential equation.

**Reply by tranqui** [reply]
I think I understand a bit better now. I doubt there is a clean way to implement this kind of boundary condition purely at the level of the u'', as it affects the other differential operators equally. As you say, you'd need some knowledge of the underlying PDE to "undo" ignorance of this BC in the other operators. I can't see a way of doing it without switching to a higher-order stencil unfortunately, which may be too unwieldy to be worthwhile.
